app/models/issuance.rb

```diff
@@ -0,0 +1,124 @@
+=begin
+  create_table "gift_cards", force: :cascade do |t|
+    t.integer "certificate"
+    t.datetime "expiration_date"
+    t.integer "registrations_available"
+    t.string "associated_product"
+    t.decimal "certificate_value"
+    t.string "gl_code"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+  end
+
+  create_table "issuances", force: :cascade do |t|
+    t.string "status"
+    t.integer "initiator_id"
+    t.decimal "card_amount"
+    t.integer "quantity"
+    t.datetime "begin_use_date"
+    t.datetime "end_use_date"
+    t.datetime "expiration_date"
+    t.integer "gift_card_type_id"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.text "allocated_certificates"
+    t.string "numbering"
+  end
+
+=end
+
+class Issuance < ApplicationRecord
+  include AASM
+  include HasNumbering
+
+  belongs_to :creator, class_name: "Person"
+  belongs_to :issuer, class_name: "Person", optional: true
+  belongs_to :gift_card_type
+  has_many :gift_cards
+
+  validates :card_amount, numericality: true 
+  validates :quantity, numericality: { only_integer: true }
+
+  aasm column: :status do
+    state :configure, initial: true
+    state :preview
+    state :issued
+
+    event :preview do
+      transitions from: :configure, to: :preview
+    end
+
+    event :issue do
+      transitions from: :preview, to: :issued, after: :create_gift_cards
+      transitions from: :issued, to: :issued
+    end
+  end
+
+  before_save do
+    unless issued?
+      # keep a copy of numbering for posterity if still previewing
+      self.numbering = gift_card_type.numbering
+      set_allocated_certificates
+    end
+  end
+
+  after_create do
+    preview!
```

Comment:
How can it ever be in the configure state? Can that happen on a create view and before initial save, or should this be `after_save`?

Reply:
Actually, ideally it's never in configure state. It was originally going to let people pick the base number, but Mona said we should just use the lowest available.. and actually it ensures the numbers stay low as long as possible and are all used up. I do like keeping the state so we know if the callback to make the preview data failed or got interrupted. I can use "initial" or "created" state or something if you prefer, since that's more accurate to reality now.